pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol

```diff
@@ -0,0 +1,326 @@
+            // boundary ratios.
+            lowerBoundRatio = circuitBreakerState.decodeUint(_LOWER_BOUND_RATIO_OFFSET, _BOUND_RATIO_WIDTH).decompress(
+                _BOUND_RATIO_WIDTH,
+                _MAX_BOUND_PERCENTAGE
+            );
+            upperBoundRatio = circuitBreakerState.decodeUint(_UPPER_BOUND_RATIO_OFFSET, _BOUND_RATIO_WIDTH).decompress(
+                _BOUND_RATIO_WIDTH,
+                _MAX_BOUND_PERCENTAGE
+            );
+        } else {
+            // Something has changed - either the weight of the token, or the composition of the pool, so we must
+            // retrieve the raw percentage bounds and do the full calculation. Decompress the bounds by shifting left.
+            (lowerBoundRatio, upperBoundRatio) = getBoundaryConversionRatios(
+                circuitBreakerState.decodeUint(_LOWER_BOUND_OFFSET, _BOUND_WIDTH) << _BOUND_SHIFT_BITS,
+                circuitBreakerState.decodeUint(_UPPER_BOUND_OFFSET, _BOUND_WIDTH) << _BOUND_SHIFT_BITS,
+                weightComplement
+            );
+        }
+
+        // Use the ratios retrieved (or computed) above to convert raw percentage bounds to BPT price bounds.
+        // To err in favor of tripping the breaker, round the lower bound up, and the upper bound down.
+        return (bptPrice.mulUp(lowerBoundRatio), bptPrice.mulDown(upperBoundRatio));
+    }
+
+    /**
+     * @notice Checks whether either the lower or upper circuit breakers would trip in the given pool state.
+     * @dev Compute the current BPT price from the input parameters, and compare it to the bounds to determine whether
+     * the given post-operation pool state is within the circuit breaker bounds.
+     * @param circuitBreakerState - the state corresponding to the token we are checking.
+     * @param totalSupply - the post-operation totalSupply (including protocol fees, etc.)
+     * @param normalizedWeight - the normalized weight of the token we are checking.
+     * @param upscaledBalance - the post-operation token balance (including swap fees, etc.). It must be an 18-decimal
+     * floating point number, adjusted by the scaling factor of the token.
+     * @return - boolean flags set to true if the breaker should be tripped: (lowerBoundTripped, upperBoundTripped)
+     */
+    function hasCircuitBreakerTripped(
+        bytes32 circuitBreakerState,
+        uint256 totalSupply,
+        uint256 normalizedWeight,
+        uint256 upscaledBalance
+    ) internal pure returns (bool, bool) {
+        (uint256 lowerBoundBptPrice, uint256 upperBoundBptPrice) = getCurrentCircuitBreakerBounds(
+            circuitBreakerState,
+            normalizedWeight.complement()
+        );
+        uint256 currentBptPrice = totalSupply.mulUp(normalizedWeight).divDown(upscaledBalance);
+
+        return (
+            lowerBoundBptPrice != 0 && currentBptPrice < lowerBoundBptPrice,
+            upperBoundBptPrice != 0 && currentBptPrice > upperBoundBptPrice
+        );
+    }
```

Comment:
I was originally thinking that we should just return a single boolean here for whether either of the breakers are flipped. Is the reasoning here that we only check the lower breaker if we're selling this token to the pool and vice versa?

Reply:
I guess this has the benefit that traders can push the pool back towards the desired range without having to do it in a single trade.